include/zephyr/drivers/rtc.h

```diff
@@ -0,0 +1,510 @@
+ */
+typedef int (*rtc_api_time_get)(const struct device *dev, struct tm *timeptr);
+
+/**
+ * @typedef rtc_alarm_time_set
+ * @brief API for setting RTC alarm time
+ */
+typedef int (*rtc_api_alarm_time_set)(const struct device *dev, uint16_t id, uint16_t mask,
+				      const struct tm *timeptr);
+
+/**
+ * @typedef rtc_alarm_time_get
+ * @brief API for getting RTC alarm time
+ */
+typedef int (*rtc_api_alarm_time_get)(const struct device *dev, uint16_t id, uint16_t *mask,
+				      struct tm *timeptr);
+
+/**
+ * @typedef rtc_alarm_enable
+ * @brief API for enabling RTC alarm
+ */
+typedef int (*rtc_api_alarm_enable)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_alarm_disable
+ * @brief API for disabling RTC alarm
+ */
+typedef int (*rtc_api_alarm_disable)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_alarm_is_enabled
+ * @brief API for testing if RTC alarm is enabled
+ */
+typedef int (*rtc_api_alarm_is_enabled)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_alarm_is_pending
+ * @brief API for testing if RTC alarm is pending
+ */
+typedef int (*rtc_api_alarm_is_pending)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_alarm_clear_pending
+ * @brief API for clearing RTC alarm pending status
+ */
+typedef int (*rtc_api_alarm_clear_pending)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_alarm_irq_callback_set
+ * @brief API for setting alarm IRQ callback
+ */
+typedef int (*rtc_api_alarm_irq_callback_set)(const struct device *dev,
+					      rtc_alarm_irq_callback callback, void *user_data);
+
+/**
+ * @typedef rtc_alarm_irq_enable
+ * @brief API for enabling RTC alarm interrupt
+ */
+typedef int (*rtc_api_alarm_irq_enable)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_alarm_irq_disable
+ * @brief API for disabling RTC alarm interrupt
+ */
+typedef int (*rtc_api_alarm_irq_disable)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_calibration_set
+ * @brief API for setting RTC calibration
+ */
+typedef int (*rtc_api_calibration_set)(const struct device *dev, int32_t calibration);
+
+/**
+ * @typedef rtc_calibration_get
+ * @brief API for getting RTC calibration
+ */
+typedef int (*rtc_api_calibration_get)(const struct device *dev, int32_t *calibration);
+
+/**
+ * @brief RTC driver API
+ */
+__subsystem struct rtc_driver_api {
+	rtc_api_time_set time_set;
+	rtc_api_time_get time_get;
+	rtc_api_alarm_time_set alarm_time_set;
+	rtc_api_alarm_time_get alarm_time_get;
+	rtc_api_alarm_enable alarm_enable;
+	rtc_api_alarm_disable alarm_disable;
+	rtc_api_alarm_is_enabled alarm_is_enabled;
+	rtc_api_alarm_is_pending alarm_is_pending;
+	rtc_api_alarm_clear_pending alarm_clear_pending;
+	rtc_api_alarm_irq_callback_set alarm_irq_callback_set;
+	rtc_api_alarm_irq_enable alarm_irq_enable;
+	rtc_api_alarm_irq_disable alarm_irq_disable;
+	rtc_api_calibration_set calibration_set;
+	rtc_api_calibration_get calibration_get;
+};
```

Comment:
@bjarki-trackunit this is still outstanding from the architecture group, it is needed for the already in-tree mcp7940n driver

Reply:
This is addressed within the `rtc_alarm_set_time()` API call, if any field of the RTC time is set, the alarm module must be enabled, if that is required by the hardware, if all fields of the alarm time are disabled, the alarm module is disabled. This rule produces consistent behavior whether the alarm enable bit is supported by an RTC or not.

The downside of this approach is that RTCs that support the enable bit will need to use the same read/write/write ops to disable and reenable an alarm, even if the alarm is set to the same time after being enabled, which could potentially be performed more efficiently for the RTCs that support the bit.

In conclusion: Since setting an alarm is not usually performed at a high frequency, the consistency and simplicity the current solution offers is prefered over the potentially saved read/write operations.